### PR TITLE
grpc server [3]: impl raftMsg client and snapshot send.

### PR DESCRIPTION
**Pull request by disksing**
The plan:
1. [x] update kvproto.
2. [x] remove kv/coprocessor from TCP server.
3. [x] impl  grpc server.
4. [x] impl raft grpc client.
5. [x] impl grpc snapshot.
6. [ ] replace TCP server with grpc servers.
7. [ ] add tests.

### CODE REVIEW

File: src/server/raft_client.rs [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+use tokio_core::reactor::Handle;
+use grpc::error::GrpcError;
+use kvproto::raft_serverpb::RaftMessage;
+use kvproto::tikvpb_grpc::TiKVAsyncClient;
+use kvproto::tikvpb_grpc::TiKVAsync;
+use util::worker::FutureRunnable;
+use util::{HashMap, TryInsertWith};
+use super::Result;
+
+// SendTask delivers a raft message to other store.
```
Comment (andelf) [inline src/server/raft_client.rs]: s/other/another/

Comment (andelf) [inline src/server/raft_client.rs]: or stores.

File: src/server/snap.rs [hunk 2]
```diff
@@ -92,16 +131,33 @@ fn send_snap(mgr: SnapManager, addr: SocketAddr, data: ConnData) -> Result<()> {
...
         return Err(box_err!("missing snap file: {:?}", s.path()));
     }
+
     // snapshot file has been validated when created, so no need to validate again.
 
-    let mut conn = try!(TcpStream::connect(&addr));
-    try!(conn.set_nodelay(true));
-    try!(conn.set_read_timeout(Some(Duration::from_secs(DEFAULT_READ_TIMEOUT))));
-    try!(conn.set_write_timeout(Some(Duration::from_secs(DEFAULT_WRITE_TIMEOUT))));
+    // Ideally, we should not collect all chunks here, and send an iterator to Grpc clint to save
```
Comment (andelf) [inline src/server/snap.rs]: typo clint

File: src/server/snap.rs [hunk 3]
```diff
@@ -92,16 +131,33 @@ fn send_snap(mgr: SnapManager, addr: SocketAddr, data: ConnData) -> Result<()> {
...
+    // memory. But inside grpc client, the iterator need to be move to another thread, and we can't
```
Comment (andelf) [inline src/server/snap.rs]: needs
moved

File: src/server/raft_client.rs [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fmt;
+use std::net::SocketAddr;
+use futures::{future, Future, Stream};
```
Comment (siddontang) [inline src/server/raft_client.rs]: add a blank line to separate the importing modules.

File: src/server/raft_client.rs [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+}
+
+impl fmt::Display for SendTask {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "send raft message to {:?}", self.addr)
+    }
+}
+
+struct Conn {
+    _client: TiKVAsyncClient,
```
Comment (siddontang) [inline src/server/raft_client.rs]: why not use client directly?

Comment (disksing) [inline src/server/raft_client.rs]: We create a stream for each Client to send raft messages, so we don't have to wait response, compare to unary rpc call.

Comment (siddontang) [inline src/server/raft_client.rs]: Sorry, what I mean why using `_` for the client name.

Comment (disksing) [inline src/server/raft_client.rs]: To suppress 'unused variable' warning.

**Pull request closed without merging**